src/main/java/org/broadinstitute/hellbender/utils/hmm/segmentation/HiddenMarkovModelPostProcessor.java

```diff
@@ -0,0 +1,1145 @@
+     * worth to be reported as an error or warning.
+     */
+    public static final double MAX_LOG_PROB = 1e-3;
+
+    /**
+     * Maximum reportable output quality score; higher quality scores will
+     * capped to this value.
+     */
+    public static final double MAX_QUAL_SCORE = 199.99999999999;
+
+    /**
+     * Precision of Phred-scale scores
+     */
+    public static final double PHRED_SCORE_PRECISION = 0.1;
+
+    /**
+     * Calculating variance posterior on long segments is time consuming; we assume hidden states become
+     * independent if their separation is larger than the following value
+     */
+    public static final double INDEPENDENT_TARGETS_SEPARATION_THRESHOLD = 1000;
+
+    /**
+     * Threshold used to determine best way to calculate log(1- exp(a))
+     * based on https://cran.r-project.org/web/packages/Rmpfr/vignettes/log1mexp-note.pdf
+     */
+    private static final double LN_1_M_EXP_THRESHOLD = - Math.log(2);
+
+    /**
+     * Cached value of ln(10)^-1
+     */
+    private static final double INV_LN_10 = 1.0 / Math.log(10);
+
+    private final List<S> allStates;
+    private final S referenceState;
+    private final List<S> alternativeStates;
+
+    private final List<Allele> allAlleles;
+    private final List<Allele> alternativeAlleles;
+
+    private final List<String> sampleNames;
+    private final int numSamples;
+    private final List<TargetCollection<T>> sampleTargets;
+    private final List<ForwardBackwardAlgorithm.Result<D, T, S>> sampleForwardBackwardResults;
+    private final List<List<S>> sampleBestPaths;
+
+    /**
+     * An intermediate result of segment calling -- will be used for VCF creation if an external
+     * segments file is not given
+     */
+    private Map<String, List<HiddenStateSegment<S, T>>> allSegmentsBySampleName = null;
+
+    /**
+     * This "master" target collection is the lexicographically sorted union of all targets from all samples
+     */
+    private final TargetCollection<T> masterTargetCollection;
+
+    /**
+     * Public constructor.
+     *
+     * @param sampleNames list of sample names
+     * @param sampleTargets list of target collection for each sample
+     * @param sampleForwardBackwardResults list of forward-backward result for each sample
+     * @param sampleBestPaths list of best hidden state sequence call for each sample
+     * @param referenceState the reference state
+     */
+    public HiddenMarkovModelPostProcessor(
+            @Nonnull final List<String> sampleNames,
+            @Nonnull final List<TargetCollection<T>> sampleTargets,
+            @Nonnull final List<ForwardBackwardAlgorithm.Result<D, T, S>> sampleForwardBackwardResults,
+            @Nonnull final List<List<S>> sampleBestPaths,
+            @Nonnull final S referenceState) {
+
+        /* basic consistency checks on the input data */
+        numSamples = sampleNames.size();
+        if (numSamples == 0) {
```

Comment:
Since this isn't a command line program, you can replace all the `USerException.BadInput`s with `IllegalArgumentException`s, which allows you to use `Utils.validateArg`.